distributed/tests/test_scheduler.py

```diff
@@ -1870,6 +1872,9 @@ async def test_task_groups(c, s, a, b):
     assert tg.states["forgotten"] == 5
     # Ensure TaskGroup is removed once all tasks are in forgotten state
     assert tg.name not in s.task_groups
+    assert tg._start > start
+    assert tg._stop < stop
```

Comment:
This fails in CI here since `stop` is a few ms shorter than `tg._stop`. The startstops of the compute are using the clock of the workers. I assume this is somehow connected. Although I would think that the clock for all workers, scheduler and test are the same considering they share the same loop

Reply:
It looks like the start/stop times the worker reports are augmented by the the worker's `scheduler_delay` estimate

https://github.com/dask/distributed/blob/843c704d0bc2f3fe576ed5b7411ca7da2c4162a4/distributed/worker.py#L3625-L3626

which might be why things are a few milliseconds off.

One approach would be to include this scheduler delay in the `start` / `stop` times used in `test_task_groups`:

```diff
diff --git a/distributed/tests/test_scheduler.py b/distributed/tests/test_scheduler.py
index 38621770..5bfedefe 100644
--- a/distributed/tests/test_scheduler.py
+++ b/distributed/tests/test_scheduler.py
@@ -1830,14 +1830,15 @@ async def test_no_danglng_asyncio_tasks(cleanup):
     assert tasks == start


-@gen_cluster(client=True)
-async def test_task_groups(c, s, a, b):
-    start = time()
+
+@gen_cluster(client=True, nthreads=[("127.0.0.1", 1)])
+async def test_task_groups(c, s, a):
     da = pytest.importorskip("dask.array")
+    start = time() + a.scheduler_delay
     x = da.arange(100, chunks=(20,))
     y = (x + 1).persist(optimize_graph=False)
     y = await y
-    stop = time()
+    stop = time() + a.scheduler_delay

     tg = s.task_groups[x.name]
     tp = s.task_prefixes["arange"]
```

While this gets the test passing for me locally, I'll admit it doesn't feel like a great solution as the `Worker.scheduler_delay` is periodically updated during the worker's heartbeat

Reply:
That makes much more sense than differing clocks :)

Once could instead force the value to be zero, e.g.

```diff
diff --git a/distributed/tests/test_scheduler.py b/distributed/tests/test_scheduler.py
index 38621770..f21bde7a 100644
--- a/distributed/tests/test_scheduler.py
+++ b/distributed/tests/test_scheduler.py
@@ -1829,8 +1829,17 @@ async def test_no_danglng_asyncio_tasks(cleanup):
     tasks = asyncio.all_tasks()
     assert tasks == start

+class NoDelayWorker(Worker):
+    @property
+    def scheduler_delay(self):
+        return 0

-@gen_cluster(client=True)
+    @scheduler_delay.setter
+    def scheduler_delay(self, value):
+        pass
+
+
+@gen_cluster(client=True, Worker=NoDelayWorker)
 async def test_task_groups(c, s, a, b):
     start = time()
     da = pytest.importorskip("dask.array")
```